### DISCUSSION TITLE
How to properly submit a PR adding public API and using that in plugins?

### DISCUSSION THREAD

**Discussion by lightmare** [babel/babel#13846] [Q&A] [answered]
Every once in a while I want to directly construct AST node for `void 0` or `(0, expr)`.
So I added shortcut builders for those to babel-types.

But now if I patch existing plugins to use those builders instead of the original long-form construction, they should depend on a future version of babel-types, right? Is there a way to submit these "together" somehow, or does the public API need to be in a public version first, and then the dependants follow?

**Comment by JLHwung** [comment, marked as answer]
> they should depend on a future version of babel-types

Yes. If you are writing a new plugin, you can bump the `@babel/core` peer deps. Otherwise for existing babel plugins, we have to support both. For example, in https://github.com/babel/babel-loader/blob/7fdf6f4ca85d4f7c25ca7952b6ec51f5cb059762/src/index.js#L163 we try `Babel.loadPartialConfigAsync` and then fallback to `babel.loadPartialConfig`.

> Is there a way to submit these "together" somehow

Yes they can be submitted in a single PR. Though the PR will be likely marked as a new feature therefore it can't get merged until the next release. 

> does the public API need to be in a public version first, and then the dependants follow?

No. We don't have to wait until the public API is published. Yarn will publish all changed packages and accordingly update the dependencies spec.
